python/lsst/daf/butler/core/config.py

```diff
@@ -730,6 +763,61 @@ def dumpToFile(self, path):
         with open(path, "w") as f:
             self.dump(f)
 
+    def dumpToS3(self, bucket, key):
+        """Writes the config to a file in S3 Bucket.
+
+        Parameters
+        ----------
+        bucketname: `str`
+            Name of the Bucket into which config will be written.
+        key : `str`
+            Path to the file to use for output, relative to the bucket.
+        """
+        stream = io.StringIO()
+        self.dump(stream)
+        stream.seek(0)
+
+        s3 = boto3.client('s3')
+        s3.put_object(Bucket=bucket, Key=key, Body=stream.read())
+
+        stream.close()
+
+    def dumpToUri(self, uri):
+        """Writes the config to location pointed to by given URI.
+
+        If URI does not specify the filename, by default `butler.yaml` will be
+        used. Currently supports 's3' and 'file' URI schemes.
+
+        Parameters
+        ----------
+        uri: `str` or `ButlerURI`
+            URI of location where the Config will be written.
+        """
+        if boto3 is None:
+            raise ModuleNotFoundError(("Could not find boto3. "
+                                       "Are you sure it is installed?"))
+
+        if isinstance(uri, str):
+            uri = ButlerURI(uri)
+
+        if uri.scheme == 'file':
+            if os.path.isdir(uri.path):
+                uri = ButlerURI(os.path.join(uri.path, 'butler.yaml'))
+            self.dumpToFile(uri.path)
+        elif uri.scheme == 's3':
+            # we guesstimate if dir or a file. Paths that *end* on strings with
+            # dots `[A-Za-z0-9].[A-Za-z0-9]` are assumed to be files.
+            isDirExpression = re.compile("\w*\.\w*$")
```

Comment:
To fix flake8 here you need to use a raw string:
```python
= re.compile(r"\w*\.\w*$")
```

Reply:
I've made the changes asked here (although I'm not sure if that's what you had in mind). The regex is kicked out and posixpath is used now and I have added `updateFile` and `defaultFileName` arguments that by default will continue to do the things the way they were done so far (assuming that the config file name is `butler.yaml` if the given path is not a file) but add the ability to overwrite the default filename to something else. 

I was not sure if you wanted the option to change the default assumed filename or not but it seemed more natural.